pcsx2/USB/usb-python2/btools/usb-python2-btools.cpp

```diff
@@ -0,0 +1,174 @@
+#include "PrecompiledHeader.h"
+#include "USB/USB.h"
+#include "usb-python2-btools.h"
+
+#include <algorithm>
+#include <chrono>
+#include <bemanitools/ddrio.h>
+
+namespace usb_python2
+{
+	namespace btools
+	{
+		struct shim_ctx
+		{
+			HANDLE barrier;
+			int (*proc)(void*);
+			void* ctx;
+		};
+
+		static unsigned int crt_thread_shim(void* outer_ctx)
+		{
+			struct shim_ctx* sctx = (struct shim_ctx*)outer_ctx;
+			int (*proc)(void*);
+			void* inner_ctx;
+
+			proc = sctx->proc;
+			inner_ctx = sctx->ctx;
+
+			SetEvent(sctx->barrier);
+
+			return proc(inner_ctx);
+		}
+
+		int BToolsInput::crt_thread_create(
+			int (*proc)(void*), void* ctx, uint32_t stack_sz, unsigned int priority)
+		{
+
+			Console.WriteLn("crt_thread_create");
+
+			struct shim_ctx sctx;
+			uintptr_t thread_id;
+
+			sctx.barrier = CreateEvent(NULL, TRUE, FALSE, NULL);
+			sctx.proc = proc;
+			sctx.ctx = ctx;
+
+			thread_id = _beginthreadex(NULL, stack_sz, crt_thread_shim, &sctx, 0, NULL);
+
+			WaitForSingleObject(sctx.barrier, INFINITE);
+			CloseHandle(sctx.barrier);
+
+			return (int)thread_id;
+		}
+
+		void BToolsInput::crt_thread_destroy(int thread_id)
+		{
+			Console.WriteLn("crt_thread_destroy");
+			CloseHandle((HANDLE)(uintptr_t)thread_id);
+		}
+
+		void BToolsInput::crt_thread_join(int thread_id, int* result)
+		{
+			Console.WriteLn("crt_thread_join");
+
+			WaitForSingleObject((HANDLE)(uintptr_t)thread_id, INFINITE);
+
+			if (result)
+			{
+				GetExitCodeThread((HANDLE)(uintptr_t)thread_id, (DWORD*)result);
+			}
+		}
+
+		void BToolsInput::InterruptReaderThread(void* ptr)
+		{
+			BToolsInput* dev = static_cast<BToolsInput*>(ptr);
+			dev->isInterruptReaderThreadRunning = true;
+
+			while (true)
+			{
+				//BTools API claims that this method will sleep/prevent banging.
+				//so we can safely loop around it.
+				if (dev->m_ddr_io_read_pad)
```

Comment:
I feel like this could be checked before the while loop so that it's only ever checked once instead of once every loop.